Solved Crash from api request answer

edwin-pixel-1 opened the pull request:
- Added Support to manage the error messages from API Request.
- Added Debug class to manage all debug messages of this App
- Changed Debug logs in SubscribersListFragment class to the new class for debug messages
- Removed debugLog method from Utils class. Replaced by Debug class methods.
- Added debug message in onFailure method to show ErrorMessages.
- Added validation in onCancelRequest method in SubscribersListInteractorImpl class and SubscriberDetailInteractor class

> Note that this branch was merged with `Resolution-Crash` branch which was merged with `master` branch

The below trello cards are solved:
https://trello.com/c/phV19aZo/18-crash-when-github-service-send-bad-answer

edwin-pixel-1 committed:
Saved Instance State in Subscriberlist Fragment

- Added support to use savedInstanceState in SubscribersListFragment class
  to avoid recreate everything on change configuration.

- Added Helper class to manage each errors from the Api requests

edwin-pixel-1 committed:
Removed Debug Logs messages

edwin-pixel-1 committed:
Hide top bar in Splash Activity

edwin-pixel-1 committed:
Removed unused method in SubscriberViewHolder class

edwin-pixel-1 committed:
Added support to manage Error messages

- Added Support to manage messages from API in SubscriberDetail MVP
- Added support to Save instance state in SubscriberDetailFragment class
- Added Debug class to manage all debug messages of this App

edwin-pixel-1 committed:
Tweak some classes of SubscribersList MVP

- Changed Debug logs in SubscribersListFragment class to the new class for debug messages

- Added debug message in onFailure method to show ErrorMessages.

- Added validation in onCancelRequest method in
  SubscribersListInteractorImpl class

- Removed debugLog method from Utils class. Replaced by Debug class
  methods.

edwin-pixel-1 committed:
Merge branch 'Resolution-Crash' into Resolution-Crash-Api

lucaslabs commented on app/src/main/java/com/github/globant/githubsubscribers/commons/utils/Debug.java:
```diff
@@ -0,0 +1,27 @@
+import android.util.Log;
+
+/**
+ * @author edwin.cobos
```
Add javaDoc.

lucaslabs commented on app/src/main/java/com/github/globant/githubsubscribers/commons/utils/ErrorMessagesHelper.java:
```diff
@@ -0,0 +1,44 @@
+    /**
+     * List of types errors
+     */
+    public enum TypeError {
```
Nice!

lucaslabs commented on app/src/main/java/com/github/globant/githubsubscribers/subscribersdetail/presenter/SubscriberDetailPresenterImpl.java:
```diff
@@ -31,23 +32,33 @@ public void onFinishedUser(User userItem) {
+        if (messageId > 0 && view != null) {
+            view.showUserError(messageId);
+        }
+        Debug.e(this.getClass().getEnclosingMethod().getName() + ": " + errorMessage);
```
To do this the Android-way:
1. Create a constant `TAG = this.getClass().getSimpleName()`.
2. Use the TAG on the debug methods.

lucaslabs commented on app/src/main/java/com/github/globant/githubsubscribers/subscribersdetail/presenter/SubscriberDetailPresenterImpl.java:
```diff
@@ -31,23 +32,33 @@ public void onFinishedUser(User userItem) {
+        if (messageId > 0 && view != null) {
+            view.showUserError(messageId);
+        }
+        Debug.e(this.getClass().getEnclosingMethod().getName() + ": " + errorMessage);
```
Use the TAG.

lucaslabs commented on app/src/main/java/com/github/globant/githubsubscribers/subscribersdetail/presenter/SubscriberDetailPresenterImpl.java:
```diff
@@ -31,23 +32,33 @@ public void onFinishedUser(User userItem) {
     @Override
     public void onDestroy() {
         view = null;
+        interactor.onCancelRequestUser();
```
Nice!

lucaslabs commented on app/src/main/java/com/github/globant/githubsubscribers/subscribersdetail/view/SubscriberDetailFragment.java:
```diff
@@ -42,14 +43,19 @@
         loadArgs();
-        repositoriesAdapter = new RepositoryAdapter();
+        if (savedInstanceState == null) {
+            setRetainInstance(true);
```
Which is the purpose of retaining the instance of the fragment?

edwin-pixel-1 replied:
The purpose of this is retain each attribute/field of the fragment to reuse it everytime that a new fragment is created. Actually, this is to reuse mainly the MVP pattern of this fragment (View, Presenter and interactor).

lucaslabs commented on app/src/main/java/com/github/globant/githubsubscribers/subscribersdetail/view/SubscriberDetailFragment.java:
```diff
@@ -69,23 +75,26 @@ private void loadArgs() {
-        recyclerViewSubscribers.setLayoutManager(new LinearLayoutManager(getActivity()));
-        recyclerViewSubscribers.setAdapter(repositoriesAdapter);
-        repositoriesAdapter.setClickListener(this);
+        if (savedInstanceState == null) {
```
Use `if(getView() != null)` instead.

lucaslabs commented on app/src/main/java/com/github/globant/githubsubscribers/subscribersdetail/view/SubscriberDetailFragment.java:
```diff
@@ -69,23 +75,26 @@ private void loadArgs() {
+            recyclerViewSubscribers.setLayoutManager(new LinearLayoutManager(getActivity()));
+            recyclerViewSubscribers.setAdapter(repositoriesAdapter);
+            repositoriesAdapter.setClickListener(this);
+        } else {
```
No need for this else.

lucaslabs commented on app/src/main/java/com/github/globant/githubsubscribers/subscriberslist/interactor/SubscribersListInteractorImpl.java:
```diff
@@ -19,25 +19,41 @@
+                    try {
+                        listener.onFailure(response.errorBody().string(), ErrorMessagesHelper.TypeError.BAD_ANSWER);
+                    } catch (IOException e) {
+                        e.printStackTrace();
```
Use your debug class to log this error.

daniellemariani replied:
+1

lucaslabs commented on app/src/main/java/com/github/globant/githubsubscribers/subscriberslist/view/SubscribersListFragment.java:
```diff
@@ -31,26 +31,34 @@
-        recyclerViewSubscribers.setLayoutManager(new LinearLayoutManager(getActivity()));
-        recyclerViewSubscribers.setAdapter(subscribersAdapter);
-        subscribersAdapter.setClickListener(this);
+        if (savedInstanceState == null) {
```
Use `if(getView() != null)` instead.

lucaslabs commented on app/src/main/java/com/github/globant/githubsubscribers/subscriberslist/view/SubscribersListFragment.java:
```diff
@@ -31,26 +31,34 @@
+            recyclerViewSubscribers.setLayoutManager(new LinearLayoutManager(getActivity()));
+            recyclerViewSubscribers.setAdapter(subscribersAdapter);
+            subscribersAdapter.setClickListener(this);
+        } else {
```
No need for this else.

lucaslabs commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,7 +1,9 @@
     <string name="app_name">Github Subscribers</string>
     <string name="splash_title">GITHUB SUBSCRIBERS</string>
-    <string name="api_client_error">Connection problems</string>
+    <string name="api_client_error_no_connection">Connection problems</string>
```
If these are the error message the user will see, please be user-friendly, let's say, `We are having connections issues, please try again later.`

daniellemariani commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,7 +1,9 @@
-    <string name="api_client_error">Connection problems</string>
+    <string name="api_client_error_no_connection">Connection problems</string>
+    <string name="api_client_error_bad_answer">Server not responding</string>
+    <string name="api_client_error_request_cancel">Request cancel</string>
```
Should it be **Request Cancelled**, shouldn't it?

edwin-pixel-1 committed:
FIXED SUGGESTIONS

- Added TAG constant to use into Debug log

- Changed String messages to be more friendly

edwin-pixel-1 committed:
Merge branch 'master' into Resolution-Crash-Api

edwin-pixel-1 committed:
Replaced debug exception message to use Debug class

- Replaced exception message to print it using th new Debug class in
  SubscribersListInteractorImpl class.

juliocatano commented:
LG

lucaslabs commented:
LG
